dataplane/internal/engine/port.go

```diff
@@ -0,0 +1,282 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"context"
+	"fmt"
+
+	fwdpb "github.com/openconfig/lemming/proto/forwarding"
+)
+
+// CreateExternalPort creates an external port (connected to other devices).
+// An external port will write some packets to the corresponding local ports (ARP, ICMP, etc).
+func CreateExternalPort(ctx context.Context, c fwdpb.ServiceClient, name string) error {
+	port := &fwdpb.PortCreateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		Port: &fwdpb.PortDesc{
+			PortType: fwdpb.PortType_PORT_TYPE_KERNEL,
+			Port: &fwdpb.PortDesc_Kernel{
+				Kernel: &fwdpb.KernelPortDesc{
+					DeviceName: name,
+				},
+			},
+		},
+	}
+	if _, err := c.PortCreate(ctx, port); err != nil {
+		return err
+	}
+
+	if err := setupPuntRules(ctx, c, name); err != nil {
+		return err
+	}
+
+	update := &fwdpb.PortUpdateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		PortId:    &fwdpb.PortId{ObjectId: &fwdpb.ObjectId{Id: name}},
+		Update: &fwdpb.PortUpdateDesc{
+			Port: &fwdpb.PortUpdateDesc_Kernel{
+				Kernel: &fwdpb.KernelPortUpdateDesc{
+					Inputs: []*fwdpb.ActionDesc{{ // Check EtherType punt rules.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: puntEtherTypeTable(name)}},
+							},
+						},
+					}, { // Check IP Proto punt rules.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: puntIPProtoType(name)}},
+							},
+						},
+					}, { // Lookup in FIB.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: fibSelectorTable}},
+							},
+						},
+					}},
+					Outputs: []*fwdpb.ActionDesc{{ // Lookup in the to port to src MAC table.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: srcMACTable}},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+	if _, err := c.PortUpdate(ctx, update); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateLocalPort creates an local (ie TAP) port.
+func CreateLocalPort(ctx context.Context, c fwdpb.ServiceClient, name string) error {
+	port := &fwdpb.PortCreateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		Port: &fwdpb.PortDesc{
+			PortType: fwdpb.PortType_PORT_TYPE_KERNEL,
+			Port: &fwdpb.PortDesc_Kernel{
+				Kernel: &fwdpb.KernelPortDesc{
+					DeviceName: name,
```

Comment:
does this `name` need to be converted to a TAP name or no? Either way should this be documented in the docstring?